components/style/values/specified/mod.rs

```diff
@@ -655,21 +654,15 @@ impl Shadow {
             if !lengths_parsed {
                 if let Ok(value) = input.try(|i| Length::parse(context, i)) {
                     lengths[0] = value;
-                    let mut length_parsed_count = 1;
-                    while length_parsed_count < length_count {
-                        if let Ok(value) = input.try(|i| Length::parse(context, i)) {
-                            lengths[length_parsed_count] = value
-                        } else {
-                            break
+                    lengths[1] = try!(Length::parse(context, input));
+                    if let Ok(value) = input.try(|i| Length::parse_non_negative(i)) {
+                        lengths[2] = value;
+                        if !disable_spread_and_inset {
```

Comment:
I think you can remove the extra check of `disable_spread_and_inset` at the end of this function returning `Shadow`, since `lengths[3]` would always be `Length::zero()` if that option is `true`. Maybe you can add a `debug_assert!` to test that invariant.